### PR TITLE
add '%' command to visual graph to search through disassembly and nav…

### PR DESCRIPTION
**Pull request by lasek0**
add '%' command in visual graph to search in disassemble and navigate to it. works with 'agfv' command

- [x] Mark this if you consider it ready to merge
- [ ] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

**Description**

added search in visual graph, possible to search via assembly code, commens, variable names, etc.
in general this works as pdr~sentence amd read an address of that and s addr to navigate graph to found line

**Copilot**

### CODE REVIEW

File: libr/core/agraph.c [hunk 1]
```diff
@@ -4078,6 +4079,109 @@ static void visual_offset(RAGraph *g, RCore *core) {
...
+		}
+		if (c == 'q') break;
+		agraph_refresh (r_cons_singleton ()->event_data);
+	}
+
+	if (core->cons->line->contents != NULL) {
+		free (core->cons->line->contents);
+	}
+	core->cons->line->contents = NULL;
+	r_core_cmd0 (core, "e scr.highlight=");
```
Comment (trufae) [inline libr/core/agraph.c:4178]:
```suggestion
	r_config_set (core->config, "scr.highlight", "");
```

File: libr/core/agraph.c [hunk 2]
```diff
@@ -4078,6 +4079,109 @@ static void visual_offset(RAGraph *g, RCore *core) {
...
+					offset = offset_max-2;
+				}
+				goto find_next;
+			}
+		}
+		if (c == 'q') break;
+		agraph_refresh (r_cons_singleton ()->event_data);
+	}
+
+	if (core->cons->line->contents != NULL) {
```
Comment (trufae) [inline libr/core/agraph.c:4174]: free already checks for null

```suggestion
```

File: libr/core/agraph.c [hunk 3]
```diff
@@ -4078,6 +4079,109 @@ static void visual_offset(RAGraph *g, RCore *core) {
...
+		free (core->cons->line->contents);
```
Comment (trufae) [inline libr/core/agraph.c:4175]:
```suggestion
	free (core->cons->line->contents);
```

File: libr/core/agraph.c [hunk 4]
```diff
@@ -4078,6 +4079,109 @@ static void visual_offset(RAGraph *g, RCore *core) {
...
+	}
```
Comment (trufae) [inline libr/core/agraph.c:4176]:
```suggestion
```

File: libr/core/agraph.c [hunk 5]
```diff
@@ -4078,6 +4079,109 @@ static void visual_offset(RAGraph *g, RCore *core) {
...
+		} else {
+			printf ("Sentence \"%s\", not found. Press 'q' for quit, any key other to conitnue", buf);
+		}
+		fflush (stdout);
+
+		free (line);
+
+		char c = r_cons_readchar ();
+		if (addr > 0) {
+			if (c == 'n') goto find_next;
```
Comment (trufae) [inline libr/core/agraph.c:4161]: missing braces {
}
```suggestion
			if (c == 'n') {
				goto find_next;
			}
```

File: libr/core/agraph.c [hunk 6]
```diff
@@ -4078,6 +4079,109 @@ static void visual_offset(RAGraph *g, RCore *core) {
...
+			if (c == 'N') {
+				offset -= 2;
+				if (offset < 1) {
+					offset = offset_max-2;
+				}
+				goto find_next;
+			}
+		}
+		if (c == 'q') break;
```
Comment (trufae) [inline libr/core/agraph.c:4170]:
```suggestion
		if (c == 'q') {
			break;
		}
```

File: libr/core/agraph.c [hunk 7]
```diff
@@ -4078,6 +4079,109 @@ static void visual_offset(RAGraph *g, RCore *core) {
...
+
+		free (line);
+
+		char c = r_cons_readchar ();
+		if (addr > 0) {
+			if (c == 'n') goto find_next;
+			if (c == 'N') {
+				offset -= 2;
+				if (offset < 1) {
+					offset = offset_max-2;
```
Comment (trufae) [inline libr/core/agraph.c:4165]:
```suggestion
					offset = offset_max - 2;
```

File: libr/core/agraph.c [hunk 8]
```diff
@@ -4078,6 +4079,109 @@ static void visual_offset(RAGraph *g, RCore *core) {
...
+
+		r_config_set_b (core->config, "asm.offset", asm_offset);
+		r_config_set_b (core->config, "asm.lines", asm_lines);
+
+		agraph_refresh (r_cons_singleton ()->event_data);
+
+		r_cons_clear_line (0);
+		printf (Color_RESET);
+		if (addr > 0) {
+			printf ("Found (%d/%d) \"%s\". Press 'n' for next, 'N' for prev, 'q' for quit, any key other to conitnue", offset+1, offset_max, line);
```
Comment (trufae) [inline libr/core/agraph.c:4151]: Use r_cons_printf or R_LOG_INFO / R_LOG_WARN, but not `printf`

File: libr/core/agraph.c [hunk 9]
```diff
@@ -4078,6 +4079,109 @@ static void visual_offset(RAGraph *g, RCore *core) {
...
+		} else {
+			printf ("Sentence \"%s\", not found. Press 'q' for quit, any key other to conitnue", buf);
```
Comment (trufae) [inline libr/core/agraph.c:4153]: same here

File: libr/core/agraph.c [hunk 10]
```diff
@@ -4078,6 +4079,109 @@ static void visual_offset(RAGraph *g, RCore *core) {
...
+		}
+		fflush (stdout);
```
Comment (trufae) [inline libr/core/agraph.c:4155]: but maybe unnecessary

```suggestion
			r_cons_flush ();
```

**Review by trufae — CHANGES_REQUESTED** [review]
Wow! that looks like a really interesting way to navigate through the graph, I love it! check out my reviews

**Comment by lasek0** [comment]
> 

Yes, I found this to be really usefull while working with graphs, so I decide to share it. Still this can be improved since searching in a graph block which is long, search does not navigate to middle of block but it navigate always to top of block instead (even if address is set [s addr] to proper line). However this is really usefull so it can be added I think as is.

**Comment by trufae** [comment]
Fix the linting. Theres a missing space before (

**Comment by lasek0** [comment]
> Fix the linting. Theres a missing space before (

sorry do not see any more missing space

**Comment by trufae** [comment]
Merged! thank you! that's a really handy feature

**Pull request merged**